lib/build-asset-map-reader.js

```diff
@@ -24,3 +24,25 @@ function readJSON(assetMap) {
 		});
 	});
 }
+
+// Retries a function that returns a promise
+// The first argument is the function
+// The second argument is an array of ms to wait in between the tries
+function retry(fn, retries) {
+	return (...params) => {
+		return fn(...params).catch(err => {
+			return new Promise((resolve, reject) => {
+				if(retries.length > 0) {
+					let backoff = retries.shift();
+					setTimeout(_ => {
+						retry(fn, retries)(...params).
+							then(resolve).
+							catch(reject);
+					}, backoff);
+				} else {
+					reject(err);
```

Comment:
…I don’t think the else branch *needs* to be inside the `Promise` constructor.

Using `util.promisify` and pulling up the “base case” (which I consider the else branch to be) leads me to this alternative:

```js
function retry(fn, retries) {
	let timeout = util.promisify(setTimeout);
	return (...params) => fn(...params).catch(err => {
		if (retries.length == 0) throw err;
		let backoff = retries.shift();
		return timeout(_ => retry(fn, retries)(...params), backoff);
	});
}
```

Reply:
Interesting. Are you sure you can `throw` inside the `catch` function? In the case of an error after the third attempt, I want to call the `catch` of the caller of `retry`, which is why I rejected the promise in that case 🤔 

Thank you for your feedback ❤️ I will look into it tomorrow in depth 👍

Reply:
Yes I am: [MDN states that](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch)

> The Promise returned by catch() is rejected if onRejected throws an error or returns a Promise which is itself rejected; otherwise, it is resolved.

(`onRejected` is the argument to `catch`)

Reply:
I just noticed that the custom variant for `setTimeout` sensibly gets rid of the callback, so that line should be more like
```js
return timeout(backoff, ...params).then(retry(fn, retries));
```
(which I’d also recommend for the manually promosified version)